Unit tests for invokedynamic with static lambdas [TG-2811]

thk123 opened the pull request:
Depends on: #1937 :white_check_mark: 

Introduces unit tests for how the `invokedynamic` instruction is translated into codet. They are not turned on, but their behaviour is moddelled on what the codet looks like for

```java
SimpleLambda simpleLambda = new SimpleLambda() {
  public void Execute() {
    callStaticMethod();
  }
};
```

Which is the intended format. I can if it would be helpful split out some of the utility changes from the test introduction, but the changes are all separated by commits and the order provides some motivation for the auxiliary changes.

Commit:
Remove methods without a implementation or usage

Commit:
Add debug information for working directory

The most common error when running unit tests is forgetting to set the
working direcotory. This causes problems when loading java classes so to
minimize this we print the working directory in the catch output.

Commit:
Extended require_goto_statements to provide meaningful errors

Commit:
Extending require utilities to be used in test

Commit:
Adding unit test for checking local lambda conversion

Verifies the structure mimics the construction of an anonymous class
that implements the interface

Commit:
Introduce checks the the function body for Execute calls the correct lambda

Commit:
Adding utility for verifying a set of statements contains a function call

thk123 commented:
@thomasspriggs @svorenova any thoughts on the design of the 3 linked commits in the descriptions would be appreciated, particularly the last one. 

(For some reason can't request your review Tom, don't know why?)

Commit:
Refactored test method to allow reuse

Commit:
Added unit test for lambda assigned

majakusber commented on unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -0,0 +1,100 @@
+
+ Module: Unit tests for converting invokedynamic instructions into codet
+
+ Author: DiffBlue Limited.
```
According to the recent update, this should be Diffblue Ltd.
https://github.com/diffblue/doc-diffblue-common/wiki/Process-for-Creating%2C-Reviewing-and-Merging-Pull-Requests

svorenova commented on unit/testing-utils/require_goto_statements.cpp:
```diff
@@ -410,3 +410,40 @@ require_goto_statements::require_entry_point_argument_assignment(
+  // TODO: Would appreciate some review comments on whether it makes the most sense:
+  // - return a vector of matching code_function_calls
+  // - return an optionalt with the first matching call
+  // - return a code_function_callt throwing an exception if none found
```
I would say the first or a combination of the first and the third option. Because we may want to be testing different calls to the same function right?

majakusber commented on unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -0,0 +1,100 @@
+        THEN("The local variable should be assigned a non-null pointer")
+        {
+          // TODO(tkiley): we don't want 11 here
+          // TODO(tkiley): This is the actual lambda which doesn't currently work
```
Can you please give an example of what we are expecting to find here, maybe a block of code from an example with anonymous class? This may be deleted or updated later when the actual instruction conversion is implemented but would be good to have a reference point here in the meantime.

thk123 replied:
Happy to though note comment - this isn't ready for review, just the three linked commits!

thk123 commented:
@svorenova My initial search found that `getcwd` is platform specific, though it seems `_getcwd()` should work on Windows? I can't easily test so I "leave this as an exercise" to someone on Windows who wants to fix this since is just a helper for running unit tests locally. 

There is a platform agnostic version in C++17 `std::filesystem::current_path();` so only 5 years or so to wait...

tautschnig commented on unit/testing-utils/load_java_class.cpp:
```diff
@@ -103,6 +107,12 @@ symbol_tablet load_java_class(
+  char temp [ PATH_MAX ];
+  std::string path = getcwd(temp,PATH_MAX) ? std::string(temp) : "";
+  INFO("Working directory: " << path);
+#endif
```
What about `get_current_working_directory` (from util/file_util.h)?

thk123 replied:
Ah wonderful - someones done the work for me - thanks!

Commit:
Extended find pointer assignment to take a regex

Commit:
Swap finding variable values to use regex

Commit:
Adding test for lambda taking array parameters

Commit:
Adding checks for parameters of the called function

Commit:
Adding tests to verify the return of the lambda wrapper method

Commit:
Adding tests for the other two returning lambdas that don't capture

Commit:
Added utiltity for getting this member components

Add documentation to find pointer assignments for a component

Commit:
Introduce tests for lambdas that are member variables

Commit:
Add natural language explanation of the test's checks

Commit:
Pull out the logic for getting the inital assignment

This will allow sharing code for testing members as well as local
variables

Commit:
Adding tests for member methods

Commit:
Adding tests for static lambdas

thk123 commented on unit/testing-utils/require_goto_statements.cpp:
```diff
@@ -141,6 +143,47 @@ require_goto_statements::find_struct_component_assignments(
   return locations;
 }
 
+require_goto_statements::pointer_assignment_locationt
```
Outstanding: document

thk123 commented on unit/testing-utils/require_goto_statements.cpp:
```diff
@@ -151,31 +194,60 @@ require_goto_statements::pointer_assignment_locationt
   const std::vector<codet> &instructions)
+{
+  INFO("Looking for symbol: " << pointer_name);
+  std::regex specialChars{R"([-[\]{}()*+?.,\^$|#\s])"};
```
special_chars

thk123 commented on unit/testing-utils/require_goto_statements.cpp:
```diff
@@ -395,3 +467,39 @@ require_goto_statements::require_entry_point_argument_assignment(
+  const std::vector<codet> &statements,
+  const irep_idt &function_call_identifier)
+{
+  // TODO: Would appreciate some review comments on whether it makes the most sense:
```
Outstanding: modify behaviour according to suggestions

mgudemann replied:
I'd say the vector is a generalized version of the other two possibilities and therefore it makes the most sense. You could even provide different functions that use the general one to provide the other functionality.

thk123 commented:
This is now ready for review in totality. I have a few comments I need to address but don't let that hold you back! I've updated the real diff in the description of this PR, along with a summary of its intended purpose.

Commit:
Use raw strings to avoid unnecessary escaping

Commit:
Modified behaviour to find function calls

thk123 commented:
Should have resolved the compile errors and my own comments

mgudemann commented on unit/testing-utils/require_goto_statements.cpp:
```diff
@@ -13,6 +13,7 @@
 #include <util/expr_iterator.h>
 #include <goto-programs/goto_functions.h>
 #include <java_bytecode/java_types.h>
+#include <iostream>
```
shouldn't be needed

mgudemann commented on unit/testing-utils/require_expr.cpp:
```diff
@@ -69,8 +70,39 @@ member_exprt require_expr::require_member(
   return symbol_expr;
 }
+
+/// Verify a given exprt is an symbol_exprt.
```
`... is a symbol_exprt.`

mgudemann commented on unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -0,0 +1,94 @@
+
+ Module: Unit tests for converting invokedynamic instructions into codet
+
+ Author: DiffBlue Ltd.
```
Diffblue

mgudemann commented on unit/testing-utils/require_goto_statements.cpp:
```diff
@@ -410,3 +410,39 @@ require_goto_statements::require_entry_point_argument_assignment(
 }
+
+/// Verify that a collection of statements contains a function call to a
+///   function whose symbol idetnfier matches the provided identifier
```
`... identifier ...`

mgudemann commented on unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -126,6 +126,21 @@ SCENARIO(
+          test_data.lambda_function_id =
+            "java::LocalLambdas.lambda$test$1:(ILjava/lang/"
+            "Object;LDummyGeneric;)V";
+          validate_lamdba_assignement(symbol_table, instructions, test_data);
```
there's a typo in that function name

thk123 replied:
Actually two!

mgudemann commented on unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -24,6 +24,7 @@ struct lambda_assignment_test_datat
   irep_idt lambda_function_id;
 
   std::vector<exprt> expected_params;
+  bool should_return_value;
```
Having been bitten in the past with uninit Booleans, I'd prefer an explicit default value

thk123 replied:
Set to `true` as more likely to produce a failing test, reminding people to set it.

mgudemann commented on unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -370,6 +370,102 @@ SCENARIO(
+            symbol_table, instructions, test_data, "paramLambda");
+        }
+        THEN(
+          "The local variable should be assigned a non-null pointer to a "
```
`... to an ...`

mgudemann commented on unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -470,3 +470,163 @@ SCENARIO(
   });
 }
+
+void validate_static_member_variable_lambda_assignment(
```
needs doxygen (I am reviewing commit-by-commit, so may be already fixed in a later commit)

thk123 replied:
And for the other one that I missed.

thk123 commented:
Should extend to test:

 - class_hierachy correctly detects that the new class extends the interface
 - check has correct fields
 - check `@class_identifier` is set correctly

As the tests pass with prototype code but don't yet work

Commit:
Correcting typo in the scenario name

Commit:
Amending path to reflect new location

Commit:
Adding check for inheritance

Commit:
Temp checkin for checking components

Commit:
Adding checks for the super class of the generated class

Commit:
Correcting typos and adding documentation to unit tests

thk123 commented:
All comments addressed, ready for final review @mgudemann @svorenova 

TG pointer bump (probably not required): diffblue/test-gen#1628

mgudemann reviewed (approved):
LGTM

majakusber commented on unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -0,0 +1,674 @@
+///    {expected_params})
+/// \param symbol_table: The loaded symbol table
+/// \param instructions: The instructions of the method that calls invokedynamic
+/// \param test_data: The parameters for the test
```
Missing doxygen for `lambda_assignment`.

thk123 replied:
Realised it was passing around a confusing structure when it was just the single non-null assignment we care about.

majakusber commented on unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -0,0 +1,674 @@
+    require_type::require_complete_class(lambda_implementor_type_symbol.type);
+
+  REQUIRE(tmp_lambda_class_type.has_base(test_data.lambda_interface));
+  REQUIRE(tmp_lambda_class_type.has_base("java::java.lang.Object"));
```
Nit: The `Object` is here because it is a superclass correct? It should appear first in the list of bases, so consider swapping the require statements for readability.

majakusber commented on unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -0,0 +1,674 @@
+    require_type::require_component(tmp_lambda_class_type, "@java.lang.Object");
+
+  const symbol_typet &super_class_type = require_type::require_symbol(
+    super_class_component.type(), "java::java.lang.Object");
```
Why do you need to be checking for the superclass component and its symbol? It is just `Object` so nothing specific for the lambda. Also, you use both `super_class` and `base_class` in the names, can you please unify it?

thk123 replied:
Good point - just removed the checks on the structure of `java.lang.Object`

majakusber commented on unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -0,0 +1,674 @@
+    require_type::require_incomplete_class(base_class_symbol.type);
+
+  require_type::require_component(super_class_type_struct, "@class_identifier");
+  // TODO verify the components of the class have been set correctly
```
This todo will be aprt of the next tickets for lambdas? Once we know how exactly the code looks right, correct?

thk123 replied:
Nope TODO no longer relevant as this is handled in a later transformation so not something we can test here. Removed the TODO

majakusber commented on unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -0,0 +1,674 @@
+        assignments, test_data.lambda_function_id);
+
+    INFO("Looking for function call of " << test_data.lambda_function_id);
+    REQUIRE(function_calls.size() == 1);
```
Lambda function can be called more than once in a method, isn't this too restrictive? Or does each lambda call get it's own object?

thk123 replied:
We're currently looking inside the generated method in the generated class - we have one such class for each lambda. I think this should be documented where the code is generated (which doesn't yet exist). Do you think something should be added to this test to make it clearer?

majakusber replied:
This was discussed offline, there indeed should be exactly one call to the lambda from the code we generate.

majakusber commented on unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -0,0 +1,674 @@
+}
+
+/// Find the assignment to the lambda and then call validate_lamdba_assignement
+/// for full validation.
```
Missing doxygen for arguments.

thk123 replied:
This was me being lazy as I wasn't sure it would help but have taken the comments from the `validate_lambda_assignment` and document the one that's not covered

majakusber commented on unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -0,0 +1,674 @@
+          test_data.lambda_interface_method_descriptor =
+            ".Execute:()Ljava/lang/Object;";
+
+          //"java::LocalLambdas.lambda$test$0:()V"
```
How does this comment relate to the code here?

thk123 replied:
Doesn't - removed

majakusber reviewed (commented):
Looks very good in general. I only have some minor comments.

Commit:
Addressing review comments

Cleaned up some logic
Removed some redundant tests relating to java.lang.Object
Added missing documentation for helper functions
Made the validate_lambda_assignment take a single assignment as the
calling functions are already checking that there is precisely one)
Addressed linting and formatting problems

thk123 commented:
@svorenova comments addressed

thk123 commented:
Fixed clang format and linting complaints
